src/services/TestTypesService.ts

```diff
@@ -1,220 +1,357 @@
-
-                if (testCodes.length > 1) {
-                    console.error("More than one testType was retrieved");
-                    throwInternalServerError();
-                }
-
-                const response: any = {
-                    id: testType.id
-                };
-
-                filterExpression.fields // Iterate through filterExpression's fields and populate them in the response
-                    .forEach((field: keyof TestCode) => {
-                        response[field] = testCodes[0][field];
-                    });
-
-                // Populating testTypeClassification that is found in testType, not testCode
-                if (filterExpression.fields.includes("testTypeClassification")) {
-                    response.testTypeClassification = testType.testTypeClassification;
-                }
-                return response;
-            });
-    }
+        return this.sort(filterdTestType);
+      })
+      .catch((error) => {
+        if (!(error instanceof HTTPError)) {
+          console.error(error);
+          error.statusCode = 500;
+          error.body = "Internal Server Error";
+        }
+
+        throw new HTTPError(error.statusCode, error.body);
+      });
+  }
 
-    /**
-     * returns null if the test-type was not found
-     * @param id
-     * @param testTypes
-     */
-    public findTestType({ id, testTypes }: { id: string; testTypes: any; }) {
-        for (const testType of testTypes) {
-            // for (let i = 0; i < testTypes.length; i++) {
-            //   const testType: any = testTypes[i];
-            if (testType.hasOwnProperty("nextTestTypesOrCategories")) {
-                const childrenTestType: any = this.findTestType({ id, testTypes: testType.nextTestTypesOrCategories });
-
-                if (childrenTestType != null) {
-                    return childrenTestType;
-                }
-            } else if (testType.hasOwnProperty("id") && testType.id === id) {
-                return testType;
-            }
+  public getTestTypesById(id: string, filterExpression: any) {
+    return this.testTypesDAO
+      .getAll()
+      .then((data) => {
+        if (data.Count === 0) {
+          throwResourceNotFound();
         }
-        return null;
-    }
 
-    public sort(testTypes: ITestType[] | any) {
-        // Pass by value
-        const testTypeArray = testTypes;
+        return data.Items;
+      })
+      .then((testTypes) => {
+        return this.findTestType({ id, testTypes });
+      })
+      .then((testType) => {
+        if (testType === null) {
+          throwResourceNotFound();
+        }
+        let testCodes: TestCode[] = testType.testCodes;
 
-        for (const testType of testTypeArray) {
-            if (testType.hasOwnProperty("nextTestTypesOrCategories")) {
-                Object.assign(testTypeArray, { nextTestTypesOrCategories: this.sort(testType.nextTestTypesOrCategories) });
-            }
+        testCodes = testCodes
```

Comment:
I think we don't need to store testCodes into a temporary variable since array methods returns new maps.
we could also move the exit condition at the top for easier readability.

```
if ( testType === null || testType.testCodes.length === 0) throwResourceNotFound()

return testType.testCodes
  .filter()
  .filter()
  ....etc...
```

Also this chaining seems to be a good candidate for functional composition, I ll have a think about it to see if we could refactor it

Reply:
as discussed was part of old code, shown by auto formatting.
Ticket raised and resolving the PR

https://jira.dvsacloud.uk/browse/CVSB-17166